### ISSUE TITLE
Generated template functions result in a memory leak

### ISSUE THREAD

**Issue by ramondeklein** [pugjs/pug#2972]
**Pug Version:** 2.0.0-rc.4
**Node Version:** 8.9.4

It seems that there is a memory leak, when the same template is used multiple times. When the following code is run multiple times, then the memory usage keeps increasing:

```js
const template = await readFileAsync(path.join(__dirname, 'test.pug'));
const templateFunc = pug.compile(template)
    const locals = {
    name: "Ramon de Klein",
    email: "mail@ramondeklein.nl"
}
const html = templateFunc(locals)
```

## Input Pug

```pug
html(xmlns= "http://www.w3.org/1999/xhtml")
  body
    p Hello from 
        a(href="mailto:"+email)=name
```

## Additional Comments

I have created a sample application, including NodeJS snapshots, that illustrate the actual problem ([ramondeklein/pug-email-templates-memleak](https://github.com/ramondeklein/pug-email-templates-memleak)).

This issue also seems to cause niftylettuce/email-templates#279.

**Comment by ForbesLindesay** [comment]
I've narrowed it down to https://github.com/ForbesLindesay/v8-memory-leak

**Comment by ForbesLindesay** [comment]
No, I submitted an issue to V8, here: https://bugs.chromium.org/p/v8/issues/detail?id=7527

You may also want to submit an issue to the node core team.

For what it's worth, not caching the compiled templates in production would be disastrous for performance even without this memory leak.

As an aside, I wonder if this leak could be caused by having a debugger attached, have you tested to see if it can be reproduced without a debugger?

**Comment by ramondeklein** [comment]
I used the [email-templates](https://github.com/niftylettuce/email-templates) package and it doesn't use caching. After upgrading my application from ancient NodeJS v0.12.8 to v8.9.4 and upgrading all packages, I noticed that my servers ran out of memory after a while.

After hooking up the debugger to the live NodeJS instances, I saw a lot of Pug-related strings popping up between the different snapshots which led me to this issue. I never really looked at the internals of [email-templates], but it's a pretty light library. I thought it would do some kind of caching, but it didn't. I ditched the package and decided to use Pug directly, cache the template function and feed the output to [juice](https://github.com/Automattic/juice) myself. That solved the issue.

Because I noticed the problem on my production servers (that run without `--inspect`), I would think that this issue isn't related to the debugger.

Thanks for looking into this issue and reporting it to the Chromium project. I'll track the issue (although it won't be a problem anymore due to caching).

**Comment by ForbesLindesay** [comment]
The issue on chromium has been closed as the bug is only visible with dev tools attached. If you can verify that this bug exists without dev tools attached, you could open an issue on the node.js issue tracker.

**Comment by hardware** [comment]
What is the solution in the meantime ? My heroku instances run out of memory during a heavy load

![](https://i.imgur.com/muK2RCs.png)

- Pug Version: **2.0.3**
- Node Version: **8.10.0**
- My pug files : https://github.com/hardware/hitmanstat.us/tree/master/views

**Comment by ForbesLindesay** [comment]
Make sure you cache any calls to `pug.compile` or `pug.compileFile`. This is super important regardless of whether there's a memory leak (and based on the response from the V8 team, I think there might not be)

**Comment by hardware** [comment]
How I can do that ? I don't make any call in my app :

https://github.com/hardware/hitmanstat.us/search?q=pug.compile
https://github.com/hardware/hitmanstat.us/search?q=pug.compileFile

`pug.compile` or `pug.compileFile` is only found here :

![](https://i.imgur.com/inywpQC.png)
![](https://i.imgur.com/ypcx1gg.png)

**Comment by ForbesLindesay** [comment]
If you call `pug.render` you should replace it with a cached call to `pug.compile` followed by calling that function.  If you call `pug.renderFile`, make sure `cache` is set to `true` in that call. If you use express to render pug, make sure `NODE_ENV=production` in your environment.
